src/operator/batch_norm-inl.h

```diff
@@ -74,48 +350,172 @@ class BatchNormOp : public Operator {
-    Tensor<xpu, 4> data;
-    Tensor<xpu, 4> out;
-    if (in_data[batchnorm::kData].ndim() == 2) {
-      Shape<4> dshape = Shape4(in_data[batchnorm::kData].shape_[0],
-                               in_data[batchnorm::kData].shape_[1], 1, 1);
-      data = in_data[batchnorm::kData].get_with_shape<xpu, 4, real_t>(dshape, s);
-      out = out_data[batchnorm::kOut].get_with_shape<xpu, 4, real_t>(dshape, s);
-    } else {
-      data = in_data[batchnorm::kData].get<xpu, 4, real_t>(s);
-      out = out_data[batchnorm::kOut].get<xpu, 4, real_t>(s);
-    }
-    Tensor<xpu, 1> slope = in_data[batchnorm::kGamma].get<xpu, 1, real_t>(s);
-    Tensor<xpu, 1> bias = in_data[batchnorm::kBeta].get<xpu, 1, real_t>(s);
-    Tensor<xpu, 1> moving_mean = aux_states[batchnorm::kMovingMean].get<xpu, 1, real_t>(s);
-    Tensor<xpu, 1> moving_var = aux_states[batchnorm::kMovingVar].get<xpu, 1, real_t>(s);
+    doForward(s, ctx, in_data, req, out_data, aux_states);
+  }
 
-    if (param_.fix_gamma) slope = 1.f;
+  /*!
+   * \brief Perform a Backward Operation, write gradient to the in_grad.
+   *
+   * \note
+   * Convention:
+   *   out_grad.size() == OperatorProperty.NumVisibleOutputs()
+   *   out_data.size() == OperatorProperty.NumOutputs()
+   * out_data can contain additional invisible returns that remembers the
+   * state carried from the Forward pass. For example mask in the dropout.
+   * The gradients are passed from visible returns in this function.
+   *
+   * \par
+   * Not all the TBlobs in the arguments will be available
+   * if you override the DeclareBackwardDependency of corresponding OperatorProperty class.
+   * Only the dependencies you declared will be available at corresponding position,
+   * the rest of the parameters are simply dummy where you will get a nullptr.
+   * You will be safe if you use the default DeclareBackwardDependency.
+   * But only declare what you need will give engine more chance for optimization.
+   *
+   * \param ctx runtime context available to this call
+   * \param out_grad the gradient value we get from of the Operator.
+   * \param in_data the array of input data.
+   * \param out_data the array of output data.
+   * \param req request types of the saving operation, can be all types.
+   * \param in_grad the array of gradient we need to write to.
+   * \param aux_states Auxiliary states of operator. Normally operator doesn't need
+   * \sa OperatorProperty, OpReqType, OpContext
+   */
+#ifdef MXNET_USE_CUDA
+  void doBackward(mshadow::Stream<gpu> *stream,
+                  const OpContext &ctx,
+                  const std::vector<TBlob> &out_grad,
+                  const std::vector<TBlob> &in_data,
+                  const std::vector<TBlob> &out_data,
+                  const std::vector<OpReqType> &req,
+                  const std::vector<TBlob> &in_grad,
+                  const std::vector<TBlob> &aux_states);
+#endif  // MXNET_USE_CUDA
+
+  void doBackward(mshadow::Stream<cpu> *stream,
```

Comment:
It is required.  One can always select not to use a GPU even when the code is compiled with GPU enabled